### PR TITLE
Replace stream().map().iterator() with Iterators#map

### PR DESCRIPTION
**Pull request by DaveCTurner**
This is a common pattern, particularly in chunked XContent encoding, but
working with streams is surprisingly expensive. This commit replaces it
with a much simpler utility that works directly on iterators.

It also simplifies the cases where we avoid creating a stream by using
`Iterators.flatMap(..., x -> Iterators.single(...))`.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
NB most of this was done mechanically with a structural find-and-replace.

File: server/src/main/java/org/elasticsearch/common/collect/Iterators.java [hunk 1]
```diff
@@ -116,6 +116,29 @@ public T next() {
...
 
+    public static <T, U> Iterator<U> map(Iterator<? extends T> input, Function<T, U> fn) {
+        if (input.hasNext()) {
+            return new MapIterator<>(input, fn);
+        } else {
+            return Collections.emptyIterator();
+        }
+    }
+
+    private record MapIterator<T, U>(Iterator<? extends T> input, Function<T, U> fn) implements Iterator<U> {
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/common/collect/Iterators.java:127]: What about `remove` and `forEachRemaining`?

Comment (thecoop) [inline server/src/main/java/org/elasticsearch/common/collect/Iterators.java:127]: Does this need to be a record? identity/hashcode don't really make sense for Functions...

Comment (thecoop) [reply server/src/main/java/org/elasticsearch/common/collect/Iterators.java:127]: Could delegate `forEachRemaining` with `input.forEachRemaining(f.compose(fn))`

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/collect/Iterators.java:127]: The default implementations are fine. We don't use them anywhere at the moment anyway, so this'd just be dead code.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/collect/Iterators.java:127]: I just followed IntelliJ's advice here, but yeah this doesn't really make sense does it

File: server/src/main/java/org/elasticsearch/common/collect/Iterators.java [hunk 2]
```diff
@@ -116,6 +116,29 @@ public T next() {
...
+        @Override
+        public boolean hasNext() {
+            return input.hasNext();
+        }
+
+        @Override
+        public U next() {
+            if (hasNext() == false) {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/collect/Iterators.java:135]: Fair point, that was just copied from `flatMap()`

File: server/src/main/java/org/elasticsearch/common/collect/Iterators.java (hunk 1 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/common/collect/Iterators.java]: This is a general-purpose implementation, so it at least should have a delegating implementation of `remove` - the default just throws

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/collect/Iterators.java]: I disagree - we can easily add such a thing if ever we need it, but we don't need it today (and I do not expect we'll need it in future either). You could make the same argument for the other utility implementations here.

**Review by thecoop — APPROVED** [review]

**Pull request merged**